Description: Require AWSTATS_ENABLE_CONFIG_DIR environmental variable
 in order to enable configdir.  Sanitize configdir to disable usage of
 external path in cgi mode (CVE-2010-4368, CVE-2010-4367).
Author: Charles Fry <debian@frogcircus.org>
Origin: upstream, http://awstats.cvs.sourceforge.net/viewvc/awstats/awstats/wwwroot/cgi-bin/awstats.pl?r1=1.958&r2=1.959
Origin: upstream, http://awstats.cvs.sourceforge.net/viewvc/awstats/awstats/wwwroot/cgi-bin/awstats.pl?r1=1.961&r2=1.962
Bug-Debian: http://bugs.debian.org/365910
Bug-Debian: http://bugs.debian.org/606263

--- a/wwwroot/cgi-bin/awstats.pl
+++ b/wwwroot/cgi-bin/awstats.pl
@@ -1716,27 +1716,28 @@
 	# Other possible directories :				"/usr/local/etc/awstats", "/etc"
 	# FHS standard, Suse package : 				"/etc/opt/awstats"
 	my $configdir         = shift;
-	my @PossibleConfigDir = ();
+	my @PossibleConfigDir = (
+			"$DIR",
+			"/etc/awstats",
+			"/usr/local/etc/awstats", "/etc",
+			"/etc/opt/awstats"
+		);
 
 	if ($configdir) {
+		# Check if configdir is outside default values.
+		my $outsidedefaultvalue=1;
+		foreach (@PossibleConfigDir) {
+			if ($_ eq $configdir) { $outsidedefaultvalue=0; last; }
+		}
 
-# If from CGI, overwriting of configdir is only possible if AWSTATS_ENABLE_CONFIG_DIR defined
-#if ($ENV{'GATEWAY_INTERFACE'} && ! $ENV{"AWSTATS_ENABLE_CONFIG_DIR"})
-#{
-#	error("Sorry, to allow overwriting of configdir parameter from an AWStats CGI usage, environment variable AWSTATS_ENABLE_CONFIG_DIR must be set to 1");
-#}
-#else
-#{
-		@PossibleConfigDir = ("$configdir");
+		# If from CGI, overwriting of configdir with a value that differs from a default value
+		# is only possible if AWSTATS_ENABLE_CONFIG_DIR defined
+		if ($ENV{'GATEWAY_INTERFACE'} && $outsidedefaultvalue && ! $ENV{"AWSTATS_ENABLE_CONFIG_DIR"})
+		{
+			error("Sorry, to allow overwriting of configdir parameter, from an AWStats CGI page, with a non default value, environment variable AWSTATS_ENABLE_CONFIG_DIR must be set to 1. For example, by adding the line 'SetEnv AWSTATS_ENABLE_CONFIG_DIR 1' in your Apache config file or into a .htaccess file.");
+		}
 
-		#}
-	}
-	else {
-		@PossibleConfigDir = (
-			"$DIR",                   "/etc/awstats",
-			"/usr/local/etc/awstats", "/etc",
-			"/etc/opt/awstats"
-		);
+		@PossibleConfigDir = ("$configdir");
 	}
 
 	# Open config file
@@ -9545,6 +9546,8 @@
 	}
 	if ( $QueryString =~ /configdir=([^&]+)/i ) {
 		$DirConfig = &Sanitize("$1");
+		$DirConfig =~ s/\\{2,}/\\/g;	# This is to clean Remote URL
+		$DirConfig =~ s/\/{2,}/\//g;	# This is to clean Remote URL
 	}
 
 	# All filters
@@ -9630,6 +9633,8 @@
 	}
 	if ( $QueryString =~ /configdir=([^&]+)/i ) {
 		$DirConfig = &Sanitize("$1");
+		$DirConfig =~ s/\\{2,}/\\/g;	# This is to clean Remote URL
+		$DirConfig =~ s/\/{2,}/\//g;	# This is to clean Remote URL
 	}
 
 	# All filters
